geo/geomfn: Implements ST_ShortestLine

abhishek20123g opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/49035

This PR implements ST_ShortestLine({geometry, geometry}) builtin
function. Which allow us to find the LineString corresponds to
the minimum distance between given two geometrical objects.

Release note (sql change): This PR implements ST_ShortestLine({geometry,
geometry}) builtin function.

abhishek20123g committed:
geo/geomfn: Implements ST_ShortestLine

Fixes https://github.com/cockroachdb/cockroach/issues/49035

This PR implements ST_ShortestLine({geometry, geometry}) builtin
function. Which allow us to find the LineString corresponds to
the minimum distance between given two geometrical objects.

Release note (sql change): This PR implements ST_ShortestLine({geometry,
geometry}) builtin function.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50292)
<!-- Reviewable:end -->

otan commented on pkg/geo/geogfn/distance.go:
```diff
@@ -159,8 +159,10 @@ type s2GeodistEdgeCrosser struct {
-	return c.EdgeCrosser.ChainCrossingSign(p.(*s2GeodistPoint).Point) != s2.DoNotCross
+func (c *s2GeodistEdgeCrosser) ChainCrossing(p geodist.Point) (bool, geodist.Point) {
+	// Returns nil for the intersection point as we don't require the intersection
+	// point for calculation of minimum distance in geography.
```
nit: replace `for calculation of minimum distance in geography.` with `as we do not have to implement ShortestLine in geography.`

abhishek20123g replied:
sure

otan reviewed (commented):
Looks neat! one comment nit

abhishek20123g commented:
Small Overview
* In case of `geometry` intersecting we have to determine the intersection point and for that, I have changed `OnIntersect()` to `Onintersect(Point)`

* For intersection case between the edge also we have to determine the intersection point as same I have changed `ChainCrossing(Point) bool` to `ChainCrossing(Point) (bool, Point)` 

* There are cases in which we are getting -0 `LineString(-0 -0, -0 -0)`. I am not sure it's issuable or not. That is we have to get rid of this.

The reason we are getting -0:
```golang
a := 0.000
b := -1.000

fmt.Println(a/b)
```
will be -0
thought in golang `-0 == 0` is true but in `PostGIS`
```
gis=# select st_equals('LINESTRING(0 0, 0 0)'::geometry, 'LINESTRING(-0 -0, -0 -0)'::geometry);
 st_equals 
-----------
 f
(1 row)

```

otan commented:
> There are cases in which we are getting -0 LineString(-0 -0, -0 -0). I am not sure it's issuable or not. That is we have to get rid of this.

Do you know when we have a negative number as a divisor?

otan commented:
```
select st_equals('LINESTRING(0 0, 0 0)'::geometry, 'LINESTRING(-0 -0, -0 -0)'::geometry);
```

this isn't equal because you can't have a linestring where both points are the same, and since that is "not valid" postgis can do whatever it wants

otan reviewed (approved):
the -0 thing seems fine to me.

abhishek20123g commented:
the negative denominator was not in `crdb`, it was present in `github.com/twpayne/go-geom/xy/lineintersector` 

https://github.com/twpayne/go-geom/blob/62c03a64717d682dd89319176fd7247944f0c518/xy/lineintersector/nonrobust_line_intersector.go#L131-L134

Thank you for the review

abhishek20123g commented:
> this isn't equal because you can't have a linestring where both points are the same, and since that is "not valid" postgis can do whatever it wants

Got it!

abhishek20123g commented:
Rebasing PR
seems like build fails is not from this PR.

otan commented:
bors r+
